File: 2dmatrixSearch.kt
class Solution {
    fun searchMatrix(matrix: Array<IntArray>, target: Int): Boolean {

        val row = matrix.size
        val column = matrix[0].size

        var i = 0
        var j = row*column - 1

        while(i<=j) {
            val mid = i+(j-i)/2
            val element = matrix[mid/column][mid%column]

            if(element>target) j=mid-1
            else if(element<target) i=mid+1
            else return true
        }
        return false

    }

File: 2D_matrix_search.py
# Time complexity : O(log(mn))
# Space Complexity : O(1)

# Approach to solve the problem - By considering the 2D matrix as 1D array/list,
# we can implement binary search to search for the given value in O(log(mn))

class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        low = 0
        n_cols = len(matrix[0])
        n_rows = len(matrix)
        high = (n_rows * n_cols) - 1

        while low <= high:
            mid = (high + low) // 2
            row = mid // n_cols
            col = mid % n_cols
            if matrix[row][col] == target:
                return True
            elif matrix[row][col] < target:
                low = mid + 1
            else:
                high = mid - 1

        